Add new game: Checkers

Jazeem opened the pull request.

lanctot commented:
Thanks!

Seems like the checkers test is causing a seg fault. This usually means there's a bad memory access (out of bounds indexing into an array). 

To find it, I recommend building in debug mode (compiler flags -O0 and possibly -g as well) and run the test using valgrind: https://stackoverflow.com/questions/5134891/how-do-i-use-valgrind-to-find-memory-leaks.

lanctot commented:
Setting the `BUILD_TYPE` environment variable to `Debug` should have all the flags you need, see here https://github.com/deepmind/open_spiel/blob/c1fbaca644aaf6c7dd82eea4f62cccd9f6469a5d/open_spiel/CMakeLists.txt#L43. So you need to rebuild from scratch: (remove the build directory entirely, then set the env var and rerun the build) then run it manually through valgrind.

Jazeem commented:
Thanks. Let me look into it!

Jazeem commented:
Hi @lanctot, something I've noticed in clobber. In this line, shouldn't the rank of the TensorView always be 3 instead of kCellStates? 

Maybe I'm missing something.

https://github.com/deepmind/open_spiel/blob/c1fbaca644aaf6c7dd82eea4f62cccd9f6469a5d/open_spiel/games/clobber.cc#L393

lanctot commented:
> Hi @lanctot, something I've noticed in clobber. In this line, shouldn't the rank of the TensorView always be 3 instead of kCellStates?

`kCellStates` is always 3: one for empty, one for black, and one for white: https://github.com/deepmind/open_spiel/blob/c1fbaca644aaf6c7dd82eea4f62cccd9f6469a5d/open_spiel/games/clobber.cc#L33.

We used `kCellStates` here to be explicit that there's one plane per cell state (empty, black, white) and also for consistency with the other games that do it this way.

lanctot commented:
Thanks for fixing. I have not looked at the code yet. Curious: how did you handle the case of multiple captures?

Jazeem commented:
https://github.com/deepmind/open_spiel/blob/c1fbaca644aaf6c7dd82eea4f62cccd9f6469a5d/open_spiel/games/clobber.cc#L393

> We used `kCellStates` here to be explicit that there's one plane per cell state (empty, white, black) and also for consistency with the other games that do it this way.

My doubt was, to do this wouldn't you have replaced `kNumPlayers + 1` with `kCellStates` inside the braces? 

Suppose if I were to add another dimension to the Tensor, a boolean to indicate whether the cell is on the boundary or not. Wouldn't that increase the rank from 3 to 4, although `kCellStates` value is 3?

Maybe I've got this jumbled up and should go through tensor_view.h

lanctot commented:
Sorry -- you are indeed correct. I had forgotten the use of TensorView, but it refers to the number of dimensions and should not be confused with the cell states. (It just happens to be 3 in this case, so it works.)

This should read TensorView<3> in clobber. (Do you want to submit a separate PR? If not, very easy to change on our side).

Yeah, using Tensorview does make encoding the observation tensors easier and less error-prone. See the one in breakthrough for example: https://github.com/deepmind/open_spiel/blob/0ffd9584c8be560f4e0f659b3ad93933d9078b48/open_spiel/games/breakthrough.cc#L326

Jazeem commented:
> This should read TensorView<3> in clobber. (Do you want to submit a separate PR? If not, very easy to change on our side).

I can create a separate PR. 

P.S. I was checking the contributor graph and noticed 1 commit against my name, even though when I click on it I can see several more. Curious to know if I'm missing any git configurations to have caused this discrepancy.

lanctot commented:
> P.S. I was checking the contributor graph and noticed 1 commit against my name, even though when I click on it I can see several more. Curious to know if I'm missing any git configurations to have caused this discrepancy.

Hmm, yeah.. I see that but I'm not really sure what the cause is. 

I think our import/export tool (copybara) squashes all the commits from one PR into a single commit when it gets imported. But there has to be more to it than that, because you also updated games.md in a separate PR. Let's see what it looks like after this next PR and Checkers. Could also be how the views get summarized/aggregated on the github side.

Jazeem commented:
> But since you've mentioned it. I've noticed an edge case I've missed for multiple jumps. I'll update the code, add test cases and push again.

I've fixed this and updated the code, test case

Jazeem commented:
Hi @lanctot, Did you get a chance to look at the code?

lanctot commented:
@tewalds will take a look. Apologies for the delay.

lanctot commented on open_spiel/games/checkers.h:
```diff
@@ -0,0 +1,134 @@
+
+// Implementation of the board game Checkers.
+// https://en.wikipedia.org/wiki/Checkers
+//
```
I noticed your implementation has rows and columns as paramters. We usually list them all here (see headers of other games)

lanctot commented on open_spiel/games/checkers.h:
```diff
@@ -0,0 +1,134 @@
+//
+// Some notes about this implementation:
+// - Drawing: 
+//     Game is drawn if no pieces have been removed in 40 moves
```
40 Seems a bit arbitrary-- is this a commonly chosen value for this game? If so maybe mention that in the comments here?

lanctot commented on open_spiel/games/checkers.h:
```diff
@@ -0,0 +1,134 @@
+
+  Player current_player_ = 0;  // Player zero (White, 'o') goes first.
+  Player outcome_ = kInvalidPlayer;
+  int multiple_jump_piece_ = 0; // Piece in the board who can do multiple jump. Represented by row * rows_ + column
```
Can you say a bit more in the comments here about how this variable is used or why it's needed?

Jazeem replied:
This variable is used to indicate that the last moved piece can make an additional jump (multiple jumps). In that case LegalActions is filtered to only represent moves available for this piece.

I'm not entirely happy with the implementation since I had to use a global variable, if you have other thoughts, do share them

lanctot replied:
Ok great. But, what I meant was: could you elaborate on that in the comment?

Also, can we please clarify what version of the rules you are using (at the top of the header file)? I wasn't aware of forced multi-captures, e.g. I don't see anything about it mentioned on https://en.wikipedia.org/wiki/Checkers. Maybe this is a common convention used in computer Checkers? If so, please clarify that in the documentation at the top of the file.

Jazeem replied:
It's mentioned in the Wiki link that
> When capturing an opponent's piece is possible, capturing is mandatory in most official rules.
That's why I added those. About the 40 move rule, I could find some mentions of it online
> http://www.flyordie.com/games/help/checkers/en/games_rules_checkers.html

tewalds commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+  const int start_column = values[1];
+  const int direction = values[2];
+  const int move_type = values[3];
+  const int captured_piece_type = values[4];
```
Ah, I see, this is why the piece types are in there. Is it better to put more details in the history and undo based on that instead of including it in the action?

Jazeem replied:
Great suggestion. Have introduced a TurnHistoryInfo struct to do this. Now actions have reduced to 1/4th.

tewalds commented on open_spiel/games/checkers_test.cc:
```diff
@@ -0,0 +1,96 @@
+void BasicSerializationTest() {
+  std::shared_ptr<const Game> game = LoadGame("checkers");
+  std::unique_ptr<State> state = game->NewInitialState();
+  std::unique_ptr<State> state2 = game->DeserializeState(state->Serialize());
```
It seems worth testing a more complicated state.

Jazeem replied:
Added one more test with a board after a few moves

tewalds commented on open_spiel/games/checkers_test.cc:
```diff
@@ -0,0 +1,96 @@
+}
+
+void BasicCheckersTests() {
+  testing::LoadGameTest("checkers");
```
Can you test different sizes as well?

Jazeem replied:
Added it

tewalds commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+    absl::StrAppend(&result, RowLabel(rows_, r));
+
+    for (int c = 0; c < columns_; c++) {
+      absl::StrAppend(&result, StateToString(BoardAt(r, c)));
```
Is it worth distinguishing between empty black/white squares, as in one that pieces can be on vs ones that are inaccessible? Maybe with dot vs space?

Jazeem replied:
I didn't understand this correctly. Can you elaborate?

tewalds commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+      SetBoard(end_row, end_column, CellState::kEmpty);
+      CellState captured_piece = OpponentState(PlayerToState(player));
+      SetBoard((start_row + end_row) / 2, (start_column + end_column) / 2, 
+        captured_piece_type == PieceType::kMan ? captured_piece : CrownState(captured_piece));
```
This wrapping is a bit odd.

Jazeem replied:
Have introduced a function here to improve readability

tewalds reviewed (changes requested):
Thanks for submitting this game! It's a great addition!

Feel free to push back on some of the more design type comments/suggestions.

It's also worth taking a look at the Google C++ style guide. You followed most of it, but in particular the 80 char line length limit needs a closer look.

Jazeem commented on open_spiel/games/checkers.h:
```diff
@@ -0,0 +1,134 @@
+namespace open_spiel {
+namespace checkers {
+
+inline constexpr int kNumPlayers = 2;
```
That's true. Removed it.

Jazeem commented on open_spiel/games/checkers.h:
```diff
@@ -0,0 +1,134 @@
+
+  Player current_player_ = 0;  // Player zero (White, 'o') goes first.
+  Player outcome_ = kInvalidPlayer;
+  int multiple_jump_piece_ = 0; // Piece in the board who can do multiple jump. Represented by row * rows_ + column
```
Have fixed these

Jazeem commented on open_spiel/games/checkers.h:
```diff
@@ -0,0 +1,134 @@
+
+  Player current_player_ = 0;  // Player zero (White, 'o') goes first.
+  Player outcome_ = kInvalidPlayer;
+  int multiple_jump_piece_ = 0; // Piece in the board who can do multiple jump. Represented by row * rows_ + column
```
Yes. This variable is used to identify a piece with a multiple jump possibility that has completed it's first jump. Hence there'll be only one such piece.

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+// Copyright 2019 DeepMind Technologies Limited
```
Fixed

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+// Index 1: Direction is diagonally up-right.
+// Index 2: Direction is diagonally down-right.
+// Index 3: Direction is diagonally down-left.
+constexpr std::array<int, kNumDirections> kDirRowOffsets = {{-1, -1, 1, 1}};
```
Keeping this as is improves readability IMO

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+namespace {
+
+// Number of rows with pieces for each player
+inline constexpr int kNumRowsWithPieces = 3;
```
Fixed

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+    case CellState::kBlack:
+      return CellState::kBlackCrowned;
+    default:
+      SpielFatalError(absl::StrCat("Invalid state"));
```
Fixed

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+  if (row == 0 && state == CellState::kWhite) {
+    state = CellState::kWhiteCrowned;
+  }
+  if (row == kDefaultRows - 1 && state == CellState::kBlack) {
```
Yes. My bad. Have fixed these instances.

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+    case CellState::kBlackCrowned:
+      return PieceType::kKing;
+    default:
+      SpielFatalError(absl::StrCat("Invalid state"));
```
Fixed

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+std::string StateToString(CellState state) {
+  switch (state) {
+    case CellState::kEmpty:
+      return ".";
```
I thought keeping it in a switch statement would improve readability

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+  switch (player) {
+    case 0:
+      return CellState::kWhite;
+    case 1:
```
Since there are 2 players and 5 cell states, I thought being explicit would be better

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+}
+
+std::string RowLabel(int rows, int row) {
+  int row_number = 1 + (rows - 1 - row);
```
True. This function was a remnant when I cloned Clobber to write checkers. I've simplified this, I can make the changes in Clobber too if needed.

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+}
+
+std::string ColumnLabel(int column) {
+  std::string label = "";
```
This was also part of the Clobber code. But I think rewriting like this could cause a memory leak, what do you think?

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+    case CellState::kBlack:
+      return stream << "Black";
+    case CellState::kWhiteCrowned:
+      return stream << "WhiteCrowned";
```
Yes, makes sense. Have renamed this.

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+      if ((row + column) % 2 == 1) {
+        if (row >= 0 && row < kNumRowsWithPieces) {
+          SetBoard(row, column, CellState::kBlack);
+        } else if (row >= (kDefaultRows - kNumRowsWithPieces)) {
```
Yes. Fixed this.

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+  // is an Empty cell. Population goes from top left to bottom right.
+  for (int row = 0; row < rows_; row++) {
+    for (int column = 0; column < columns_; column++) {
+      char state_character = board_string[1 + row * columns_ + column];
```
Changed this.

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+
+void CheckersState::SetCustomBoard(const std::string board_string) {
+  SPIEL_CHECK_EQ(rows_ * columns_, board_string.length() - 1);
+  current_player_ = board_string[0] - '0';
```
Added this

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+
+CellState CrownStateIfLastRowReached(int row, CellState state) {
+  if (row == 0 && state == CellState::kWhite) {
+    state = CellState::kWhiteCrowned;
```
Fixed it

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+
+void CheckersState::DoApplyAction(Action action) {
+  std::vector<int> values =
+      UnrankActionMixedBase(action, {rows_, columns_, kNumDirections, kNumMoveType, kNumPieceType, kNumPieceType});
```
Makes sense. Have added a function and struct

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+    current_player_ = 1 - current_player_; 
+  }
+
+  if (LegalActions().empty()) {
```
I don't think so, but it could be put inside the if block.

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+      SetBoard(start_row, start_column, CellState::kEmpty);
+      moves_without_capture_ = 0;
+
+      // Check if multiple jump is possible
```
That's fair. I've added more comments here

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+      for (Action action: moves) {
+        std::vector<int> move = UnrankActionMixedBase(action, {rows_, columns_, kNumDirections, kNumMoveType, kNumPieceType, kNumPieceType});
+        if (move[0] == end_row && move[1] == end_column && move[3] == MoveType::kCapture) {
+          moves_for_last_moved_piece.push_back(action);
```
Great suggestion. Code is simplified here now.

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+        }
+      }    
+      if (moves_for_last_moved_piece.size() > 0) {
+        multiple_jump = true;
```
Have removed the flag, so there's only a single source of truth now

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+            CellState opponent_state_crowned = CrownState(opponent_state);
+
+            if (adjacent_state == CellState::kEmpty) {
+              action_values[0] = row;                                 // Initial row value of player piece
```
Added it

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+              action_values[0] = row;                                 // Initial row value of player piece
+              action_values[1] = column;                              // Initial column value of player piece
+              action_values[2] = direction;                           // Direction of move for player piece
+              action_values[3] = MoveType::kNormal;                   // Type of move
```
To make use of in UndoAction. 

Captured piece: Needed to check if player captured a normal or crowned opponent piece
Player piece: Needed because an Undo move from the last row doesn't always mean that the piece should be uncrowned

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+
+  // If capture moves are possible, it's mandatory to play them.
+  if (!capture_move_list.empty()) {
+    if (multiple_jump_piece_ > 0) {
```
Changed this to !=-1

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+  }
+}
+
+void CheckersState::DoApplyAction(Action action) {
```
Added in the header comments

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+}
+
+std::string CheckersState::ToString() const {
+  std::string result = "";
```
Is this to optimise the code performance?

Jazeem commented on open_spiel/games/checkers.cc:
```diff
@@ -0,0 +1,546 @@
+                               true);
+
+  // Observation Tensor Representation:
+  //   Plane 0: 1's where the current player's pieces are, 0's elsewhere.
```
Crowned pieces are added in another Plane 2 & 3

lanctot commented:
Merged in https://github.com/deepmind/open_spiel/commit/808e54ef65c4f9d8597fbecf957ce6ead13fb6a1
